README.md
# lidarvegdensity
Provides functions to estimate vegetation density profiles from terrestrial LiDAR based on
the methods described in:

Ashcroft MB, Gollan JR, Ramp D (2014) *Creating vegetation density profiles for a diverse range of 
  ecological habitats using terrestrial laser scanning.*
  Methods in Ecology and Evolution 5, 263–272. doi:10.1111/2041-210X.12157.

The original paper included R code (as an online appendix) for two functions: the first to
derive a 'null' or 'reference' model from a scan performed in a closed room; and the second 
to process a vegetation site scan with the null model used to correct for non-returned pulses
when calculating vegetation density within voxels.

This package provides functions for the same steps, but with sections of the code written in 
Rcpp to greatly reduce the time taken to process scans. It also provides an option to calculate 
bounds, as a [Jeffrey's interval](https://en.wikipedia.org/wiki/Binomial_proportion_confidence_interval#Jeffreys_interval),
around the proportion of reflected rays in each voxel.

## Installation

It is best to use R version 4.0.0 or newer.

If you are a Windows user, make sure that you have Rtools installed on your system to build the C++ code included in the package. 
See https://cran.r-project.org/bin/windows/Rtools/ for more information.

To download and build the package, run the following code from the R or RStudio console...

```
# Uncomment and run the next line if you don't have the devtools package installed
# install.packages("devtools")

devtools::install_github("mbedward/lidarvegdensity")

```


NEWS.md
# NEWS lidarvegdensity

Version 0.1.8 2020-10-29

  * Fixed problem where matrix of ground elevations was being read incorrectly due
    to different X-Y index order of result arrays and the elevation raster. The
    conventional X-Y index order used by the raster package is now applied to all
    inputs and outputs.

  * Function get_prop_reflected now checks for scan points outside the vertical
    limits of the null reference model. Also added arguments to specify whether
    the function should ignore scan points outside the horizontal or vertical
    range, or stop with an error. Default behaviour is to ignore points outside
    the horizontal range and stop with an error for points outside the vertical
    range.

  * Function get_prop_reflected now accepts either a RasterLayer object or a
    matrix (in raster cell order) for the ground argument.

  * The named list returned by function get_max_returns now has a class attribute
    'TLSNullModel'.

  * The named list returned by function get_prop_reflected now has a class attribute
    'TLSResult'.

  * Added function res_to_brick to create a RasterBrick object from one of the
    arrays in the result object returned by get_prop_reflected. By default, the
    function creates a brick for the 'preflected' array. A brick object is useful
    for plotting and GIS-style calculations.


R/public_helpers.R
#' Iteratively replace NA values in a numeric matrix
#'
#' This function iteratively replaces missing values in a numeric matrix with
#' the mean of neighbouring cell values. Both orthogonal and diagonal neighbours
#' are included when calculating the mean value. The function is used by
#' \code{get_prop_reflected} to remove missing values from ground elevation
#' matrices.
#'
#' @param M A numeric matrix possibly containing missing values
#'
#' @return A numeric matrix with missing values replaced. If all values were
#'   missing in the input matrix, a warning message is issues and the matrix is
#'   returned unchanged.
#'
#' @export
#'
nibble_matrix <- function(M) {
  stopifnot(is.matrix(M))
  di <- -1:1

  clamp <- function(j) max(1,min(j)):min(ncol(M),max(j))

  na.inds <- which(is.na(M), arr.ind = TRUE)
  if (nrow(na.inds) == nrow(M)) {
    warning("All values are missing in input matrix")
    return(M)
  }

  while (nrow(na.inds) > 0) {
    nfixed <- 0
    for (ina in 1:nrow(na.inds)) {
      ir <- na.inds[ina,1]
      ic <- na.inds[ina,2]
      kr <- clamp(ir + di)
      kc <- clamp(ic + di)
      k.inds <- as.matrix(expand.grid(kr, kc))
      val <- mean(M[k.inds], na.rm = TRUE)
      if (!is.na(val)) {
        M[ir,ic] <- val
        nfixed <- nfixed + 1
      }
    }

    # If we failed to fix any of the NA cells on this iteration
    # it's time to give up
    if (nfixed == 0) break
    else na.inds <- which(is.na(M), arr.ind = TRUE)
  }

  M
}


#' Create a raster brick from \code{get_prop_reflected} results
#'
#' This function takes a result list object returned by
#' \code{\link{get_prop_reflected}} and creates a \code{RasterBrick} object
#' where the cell values in each layer represent the proportion of rays returned
#' for each voxel column.
#'
#' @param res A \code{TLSResult} object as returned by \code{get_prop_reflected}.
#'
#' @param elem The result element to convert to a \code{RasterBrick}. One of:
#'   'preflected' (default), 'nreflected', 'nthrough', 'nblocked', 'lower', 'upper'.
#'   An abbreviated name will work if unique. The 'lower' and 'upper' options only
#'   apply if the result object contains arrays for lower and upper bounds of the
#'   proportion of reflected rays.
#'
#' @export
#'
res_to_brick <- function(res,
                         elem = c('preflected', 'nreflected', 'nthrough', 'nblocked',
                                  'lower', 'upper')) {

  if (!inherits(res, "TLSResult")) {
    stop("Expected a list of class 'TLSResult' as returned by get_prop_reflected")
  }

  elem = match.arg(elem)

  if (elem == 'lower' || elem == 'upper') {
    ii <- grepl("^bound_", names(res))
    if (sum(ii) != 2) {
      stop("This result object does not have arrays for lower and upper bounds")
    }

    bounds.names <- names(res)[ii]

    # Assume bounds objects are listed in order lower, upper
    if (elem == 'lower') {
      ar <- res[[bounds.names[1]]]
    } else {
      ar <- res[[bounds.names[2]]]
    }
  } else { # something other than bounds
    ar <- res[[elem]]
  }

  b <- raster::brick(ar,
                     xmn = res$xylims[1], xmx = res$xylims[2],
                     ymn = res$xylims[1], ymx = res$xylims[2])

  z <- seq(res$zlims[1], res$zlims[2], by = res$zsize)[-1]
  names(b) <- paste0("to_", ifelse(z < 0, "minus_", ""), abs(z), "m")

  b
}



src/RcppExports.cpp
// Generated by using Rcpp::compileAttributes() -> do not edit by hand
// Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#include <RcppArmadillo.h>
#include <Rcpp.h>

using namespace Rcpp;

// cpp_count_voxel_rays
List cpp_count_voxel_rays(NumericMatrix scandata, int numcols, int numheights, double xysize, double minxy, double maxxy, double zsize, double minz, double maxz, double maxdist, bool countReflected, Rcpp::Nullable<NumericMatrix> groundarg);
RcppExport SEXP _lidarvegdensity_cpp_count_voxel_rays(SEXP scandataSEXP, SEXP numcolsSEXP, SEXP numheightsSEXP, SEXP xysizeSEXP, SEXP minxySEXP, SEXP maxxySEXP, SEXP zsizeSEXP, SEXP minzSEXP, SEXP maxzSEXP, SEXP maxdistSEXP, SEXP countReflectedSEXP, SEXP groundargSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< NumericMatrix >::type scandata(scandataSEXP);
    Rcpp::traits::input_parameter< int >::type numcols(numcolsSEXP);
    Rcpp::traits::input_parameter< int >::type numheights(numheightsSEXP);
    Rcpp::traits::input_parameter< double >::type xysize(xysizeSEXP);
    Rcpp::traits::input_parameter< double >::type minxy(minxySEXP);
    Rcpp::traits::input_parameter< double >::type maxxy(maxxySEXP);
    Rcpp::traits::input_parameter< double >::type zsize(zsizeSEXP);
    Rcpp::traits::input_parameter< double >::type minz(minzSEXP);
    Rcpp::traits::input_parameter< double >::type maxz(maxzSEXP);
    Rcpp::traits::input_parameter< double >::type maxdist(maxdistSEXP);
    Rcpp::traits::input_parameter< bool >::type countReflected(countReflectedSEXP);
    Rcpp::traits::input_parameter< Rcpp::Nullable<NumericMatrix> >::type groundarg(groundargSEXP);
    rcpp_result_gen = Rcpp::wrap(cpp_count_voxel_rays(scandata, numcols, numheights, xysize, minxy, maxxy, zsize, minz, maxz, maxdist, countReflected, groundarg));
    return rcpp_result_gen;
END_RCPP
}

static const R_CallMethodDef CallEntries[] = {
    {"_lidarvegdensity_cpp_count_voxel_rays", (DL_FUNC) &_lidarvegdensity_cpp_count_voxel_rays, 12},
    {NULL, NULL, 0}
};

RcppExport void R_init_lidarvegdensity(DllInfo *dll) {
    R_registerRoutines(dll, NULL, CallEntries, NULL, NULL);
    R_useDynamicSymbols(dll, FALSE);
}


R/refimpl_functions.R
# R reference implementation (no Rcpp) for testing
#
R_get_max_returns <- function(scandata, xysize, zsize, numcols,
                                   zmax = 35.25, zmin = NULL) {

  xysize <- .gtzero( .single_numeric(xysize) )
  zsize <- .gtzero( .single_numeric(zsize) )
  numcols <- .gtzero( .single_numeric(numcols) )

  zmax <- .single_numeric(zmax)

  if (is.null(zmin)) {
    if (zmax > 0) zmin <- -zmax
    else stop("When zmax <= 0 you must provide a value for zmin")
  } else {
    zmin <- .single_numeric(zmin)
    if (zmin >= zmax) stop("Value of zmin must be less than zmax")
  }

  #if (numcols %% 2 == 0) {
  #  numcols <- numcols + 1
  #  message("Adjusted numcols to ", numcols)
  #}

  # total limits of the columns
  xylims <- c(-1, 1) * numcols * xysize / 2

  # max and min height in metres are defaulted to 35.25m and -35.25m
  # ensure they are multiples of zsize
  zlims.working <- c(
    (trunc(zmin / zsize) - 1) * zsize,
    (trunc(zmax / zsize) + 1) * zsize
  )
  numheights <- abs(diff(zlims.working) / zsize)

  # calculate maximum distance of points from scanner
  maxzdist <- max(abs(zlims.working))
  maxdist <- sqrt(2*xylims[2]^2 + maxzdist^2)

  # create array to store number of ray segments in each rectangular prism
  prismrays <- array(0, dim = c(numcols, numcols, numheights))

  # We are going to go through point by point, creating a laser ray and dividing it
  # into segments that are each zsize in length. This is equivalent to
  # the vertical distance between each height class.

  # iterate through each ray
  pb <- txtProgressBar(max = nrow(scandata), style = 3)
  for (rayindex in 1:nrow(scandata)) {
    if (rayindex %% 1e4 == 0) setTxtProgressBar(pb, rayindex)

    # get the ray data for this point
    xo <- scandata[rayindex,1]
    yo <- scandata[rayindex,2]
    zo <- scandata[rayindex,3]
    dreflect <- sqrt(xo^2 + yo^2 + zo^2)

    # the points between segments can be written as x,y,z = k*delta*(xo,yo,zo) where k is
    # an integer and delta is a constant that ensures each segment has length zsize
    delta <- zsize/dreflect

    # work out how many segments within range
    maxtotsegments <- trunc(maxdist / zsize) + 1
    maxxsegments <- round(xylims[2] / abs(delta * xo))
    maxysegments <- round(xylims[2] / abs(delta * yo))

    if (zo < 0)
      maxzsegments <- round(abs(zlims.working[1] / (delta * zo)))
    else
      maxzsegments <- round(abs(zlims.working[2] / (delta * zo)))

    numsegments <- min(maxtotsegments, maxxsegments, maxysegments, maxzsegments)

    # loop over segments
    for (k in 1:numsegments) {
      # creating points at midpoints (k=0.5,1.5,etc)
      xn = (k - 0.5) * delta * xo
      yn = (k - 0.5) * delta * yo
      zn = (k - 0.5) * delta * zo

      # need to know which column the midpoint is in
      xcol <- trunc((xn - xylims[1]) / xysize) + 1
      ycol <- trunc((yn - xylims[1]) / xysize) + 1

      # and need the z columns
      zcol <- trunc((zn - zlims.working[1]) / zsize) + 1

      # inc total number of rays in the prism
      prismrays[xcol, ycol, zcol] <- prismrays[xcol, ycol, zcol] + 1
    }
  }
  setTxtProgressBar(pb, nrow(scandata))
  close(pb)

  list(returns = prismrays,
       xysize = xysize, xylims = xylims,
       zsize = zsize, zlims = c(zmin, zmax),
       numcols = numcols, numheights = numheights)
}



# R reference implementation (no Rcpp) for testing
#
# ground - either a square matrix with the same number of rows and columns as
# the refdata object; or a single value. If a matrix, values are ground elevations
# relative to scanner. A single value is equivalent to a matrix of uniform values.
#
# Returns a list with meta-data elements and a 3D array where the first two
# dimensions (rows and columns) match the input refdata object and the third
# (vertical) dimension represents voxel layers from the minimum ground height
# to the maximum scan height.
#
R_get_prop_reflected <- function(scandata, refdata, ground) {

  numcols <- refdata$numcols
  xylims <- refdata$xylims
  xysize <- refdata$xysize
  zsize <- refdata$zsize

  # Check that scan points are within the bounds of the reference model
  inside.ref <- scandata[,1] >= xylims[1] &&
    scandata[,1] <= xylims[2] &&
    scandata[,2] >= xylims[1] &&
    scandata[,2] <= xylims[2]

  if (any(!inside.ref)) {
    n <- sum(!inside.ref)
    msg <- ifelse(n == 1, "point is outside", "points are outside")
    msg <- glue::glue("{n} {msg} the bounds of the reference model and will be ignored")
    warning(msg)
  }

  # Check ground elevation data
  if (is.matrix(ground)) {
    if (nrow(ground) != refdata$numcols ||
        ncol(ground) != refdata$numcols) {

      stop("The matrix of ground elevations should be square, with number of\n",
           "rows and columns equal to the 'numcols' value in refdata")
    }

  } else if (is.numeric(ground)) {
    ground <- .single_numeric(ground)
    ground <- matrix(ground, numcols, numcols)

  } else {
    stop("Argument ground should either be a square matrix with number of\n",
         "rows and columns equal to the 'numcols' value in refdata, or a\n",
         "single value for ground elevation to apply to all voxels.")
  }

  # If there are any NA values in ground, set these to the mean
  # non-NA neighbour value. If any cells have only missing neighbour values
  # set these (arbitrarily) to the min ground value and issue a warning.
  if (anyNA(ground)) {
    warn.nas <- FALSE
    di <- -1:1
    km <- matrix(0, nrow=9, ncol=2)
    clamp <- function(x) max(1,min(x)):min(numcols,max(x))
    for (ir in 1:numcols) {
      for (ic in 1:numcols) {
        x <- 0
        n <- 0
        for (kr in clamp(ir + d)) {
          for (kc in clamp(ic + d)) {
            g <- ground[kr, kc]
            if (!is.na(g)) {
              x <- x + g
              n <- n + 1
            }
          }
        }
        if (n > 0) ground[ir, ic] <- x/n
        else {
          ground[ir, ic] <- ground.minz
          warn.nas <- TRUE
        }
      }
    }

    if (warn.nas)
      warning("Ground elevation for one or more cells could not be determined.\n",
              "Cell values have been set to overall minimum value.")
  }

  # Determine the height range and number of voxel height levels.
  # We clamp the height limits to multiples of zsize.
  ground.minz <- min(ground, na.rm = TRUE)
  minz <- (trunc(ground.minz / zsize) - 1) * zsize
  maxz <- (trunc(max(scandata[,3]) / zsize) + 1) * zsize

  stopifnot(maxz > minz)
  numheights <- (maxz - minz) / zsize


  # Find the index of the height dimension in the reference data array
  # that corresponds with the ground elevation for each voxel column
  ground.minzindex <- trunc((ground - refdata$zlims[1]) / zsize) + 1

  # Z-index offset for reference model.
  ref.zoffset <- (minz - refdata$zlims[1]) / zsize

  # calculate maximum distance of points from scanner
  maxzdist <- max(abs(c(maxz, minz)))
  maxdist <- sqrt(2*xylims[2]^2 + maxzdist^2)

  #These are the arrays where the data is stored
  # create arrays to store number of ray segments in each rectangular prism

  # prismrays stores according to raw z values (minz to maxz) and is the total
  # number of rays (blocked before reaching this prism, reflected from this prism +
  # passed through this prism)

  # throughrays and returnrays are stored according to ground level (0 to maxz-minz)
  # throughrays is number that pass through the prism without reflecting back
  # reflectedrays is number that are reflected back from within this prism
  prismrays <- array(0, dim = c(numcols, numcols, numheights))
  throughrays <- reflectedrays <- prismrays

  # We are going to go through point by point, creating a laser ray and dividing it
  # into segments that are each zsize in length. This is equivalent to
  # the vertical distance between each height class.
  # For each segment we will determine if the ray is blocked (reflection occurred before
  # segment), reflected (reflection occurred in segment) or through (reflection
  # occurred after segment) and add the totals to our output arrays


  # This is the main process for the function and takes the longest. It runs through
  # every beam and finds which voxels it passed through and which voxel it recived
  # a return from. Given there are over 14 million returns it can take a long time
  # to process. There are reasonably good details about what is going on here.
  # iterate through each ray
  for (rayindex in 1:nrow(scandata)) {
    xo <- scandata[rayindex,1]
    yo <- scandata[rayindex,2]
    zo <- scandata[rayindex,3]
    dreflect <- sqrt(xo^2 + yo^2 + zo^2)

    # the points between segments can be written as x,y,z = k*delta*(xo,yo,zo) where k is
    # an integer and delta is a constant that ensures each segment has length zsize
    delta <- zsize / dreflect

    # work out how many segments within range
    maxtotsegments <- trunc(maxdist / zsize) + 1
    maxxsegments <- round(xylims[2] / abs(delta * xo))
    maxysegments <- round(xylims[2] / abs(delta * yo))

    if (zo < 0)
      maxzsegments <- round(abs(minz / (delta * zo)))
    else
      maxzsegments <- round(abs(maxz / (delta * zo)))

    numsegments <- min(maxtotsegments, maxxsegments, maxysegments, maxzsegments)

    # work out the segment the reflection is in
    reflectedsegment <- trunc(dreflect / zsize) + 1


    # loop over segments
    for (k in 1:numsegments) {
      # creating points at midpoints (k=0.5,1.5,etc)
      xn = (k - 0.5) * delta * xo
      yn = (k - 0.5) * delta * yo
      zn = (k - 0.5) * delta * zo

      # need to know which column the midpoint is in
      xcol <- trunc((xn - xylims[1]) / xysize) + 1
      ycol <- trunc((yn - xylims[1]) / xysize) + 1

      # and need the z columns (raw for prismrays, adj for others)
      zcol <- trunc((zn - minz) / zsize) + 1

      znadj <- zn - ground[xcol, ycol]
      zcoladj <- trunc(znadj / zsize) + 1

      # inc total number of rays in the prism
      prismrays[xcol, ycol, zcol] <- prismrays[xcol, ycol, zcol] + 1

      # and add to reflected and through if it is above ground level
      if (znadj > 0) {
        if (k == reflectedsegment)
          reflectedrays[xcol,ycol,zcoladj] <- reflectedrays[xcol,ycol,zcoladj] + 1
        else if (k < reflectedsegment)
          throughrays[xcol,ycol,zcoladj] <- throughrays[xcol,ycol,zcoladj] + 1
        #else
        #  blockedrays[xcol,ycol,zcol] <- blockedrays[xcol,ycol,zcol] + 1
      }
    }
  }

  # We've gone through all the points provided in the scan, but some rays went
  # right through the vegetation without being returned. We need to add these rays
  # into the 'through' tally by comaprison with a null model in a closed room.
  for (xindex in 1:numcols) {
    for (yindex in 1:numcols) {
      for (zindex in 1:numheights) {
        # Check if we have less rays in the voxel than expected (because
        # some went through). If so, increase the through count.
        expected <- refdata$returns[xindex, yindex, zindex + ref.zoffset]
        if (prismrays[xindex,yindex,zindex] < expected) {
          nadd <- expected - prismrays[xindex, yindex, zindex]

          zminlimit <- (zindex-1) * zsize + minz
          zmaxlimit <- zindex * zsize + minz

          z <- zminlimit + 0.5 * (zmaxlimit - zminlimit)
          zadj <- z - ground[xindex,yindex]
          zcoladj <- trunc(zadj / zsize)  + 1
          throughrays[xindex,yindex,zcoladj] <- throughrays[xindex,yindex,zcoladj] + nadd
        }
      }
    }
  }

  # Return proportions
  preflect <- reflectedrays / (reflectedrays + throughrays)

  # Guard against NaN which occur when a voxel in both reflectedrays
  # and throughrays has a zero count
  preflect[is.nan(preflect)] <- 0

  list(preflect = preflect,
       xysize = xysize, xylims = xylims,
       zsize = zsize, zlims = c(minz, maxz),
       numcols = numcols, numheights = numheights,
       totalrays = prismrays,
       reflectedrays = reflectedrays,
       throughrays = throughrays)
}




R/voxel_functions.R
#' Estimate the potential maximum number of returns for voxels
#'
#' This function takes data from a control scan that was performed in a closed
#' room (i.e. no gaps or reflective surfaces) where all laser pulses are
#' expected to provide a return.
#'
#' @param scandata Either a three column matrix or data frame of XYZ coordinates
#'   for return positions relative to the scanner; or a \code{LAS} object as
#'   returned by \code{lidR::readLAS}.
#'
#' @param xysize The horizontal size (i.e. width in the X-Y plane) of voxels in
#'   metres.
#'
#' @param zsize The vertical size (height) of voxels in metres.
#'
#' @param numcols The integer number of voxel colums in the X and Y direction.
#'   Ashcroft et al. recommend that this be an odd value so that the scanner
#'   lies in the centre column, but this is not enforced here.
#'
#' @param zmax The maximum height of the scan (relative to the scanner) in
#'   metres. Default value is 35.25 as used by Ashcroft et al. This will usually
#'   be a positive value (uppermost voxel above scanner) but zero and negative
#'   values are also valid.
#'
#' @param zmin The minimum height of the scan (relative to the scanner) in
#'   metres. Must be less than \code{zmax}. Defaults to minus \code{zmax} if
#'   zmax is positive. If \code{zmax} is zero or negative an explicit value must
#'   be provided.
#'
#' @return A named list (object of class \code{TLSNullModel}) with the following
#'   elements:
#'   \describe{
#'   \item{returns}{An array (dimensions: Yrow, Xcol, Zlayer) of maximum
#'     expected number of rays passing through each voxel.}
#'   \item{xysize}{The horizontal voxel width.}
#'   \item{xlims}{A vector of min and max values for X and Y ordinates.}
#'   \item{zsize}{The vertical voxel width.}
#'   \item{zlims}{A vector or min and max values for Z (height).}
#'   \item{numcols}{Number of voxels in each horizontal dimension.}
#'   \item{numheights}{Number of voxels in vertical dimension.}
#'   }
#'
#' @examples
#' \dontrun{
#' # Estimate returns for regular cubic voxels with 0.25m side length,
#' # for a scan block with horizontal dimensions 60x60 voxels
#' refdata <- get_max_returns(control.scan,
#'                            xysize = 0.25, zsize = 0.25, numcols = 60,
#'                            zmax = 100)
#' }
#'
#' @export
#'
get_max_returns <- function(scandata, xysize, zsize, numcols,
                            zmax = 35.25, zmin = NULL) {

  if (inherits(scandata, "LAS")) {
    scandata <- as.matrix(scandata@data[, c("X", "Y", "Z")])

  } else if (is.matrix(scandata) || is.data.frame(scandata)) {
    if (!ncol(scandata) == 3) {
      stop("Argument scandata should be a three column matrix or data frame")
    }
    scandata <- as.matrix(scandata)
  }

  xysize <- .gtzero( .single_numeric(xysize) )
  zsize <- .gtzero( .single_numeric(zsize) )
  numcols <- .gtzero( .single_numeric(numcols) )

  xylims <- c(-1, 1) * numcols * xysize / 2
  names(xylims) <- c('min', 'max')

  zmax <- .single_numeric(zmax)

  if (is.null(zmin)) {
    if (zmax > 0) zmin <- -zmax
    else stop("When zmax <= 0 you must provide a value for zmin")
  } else {
    zmin <- .single_numeric(zmin)
    if (zmin >= zmax) stop("Value of zmin must be less than zmax")
  }

  # max and min height in metres, ensuring that they are
  # multiples of zsize
  zlims.working <- c(
    (trunc(zmin / zsize) - 1) * zsize,
    (trunc(zmax / zsize) + 1) * zsize)

  numheights <- abs(diff(zlims.working) / zsize)
  stopifnot(numheights > 0)

  # calculate maximum distance of points from scanner
  maxzdist <- max(abs(zlims.working))
  maxdist <- sqrt(2*xylims[2]^2 + maxzdist^2)

  # Call Rcpp function to count the number of rays passing through
  # each voxel
  counts <- cpp_count_voxel_rays(
    scandata,
    numcols, numheights,
    xysize, xylims[1], xylims[2],
    zsize, zlims.working[1], zlims.working[2],
    maxdist
  )

  # Reshape the returned vector into an array
  # (the cpp_count_voxel_rays function should have ordered the
  # vector appropriately)
  counts <- array(counts$total, dim = c(numcols, numcols, numheights))

  res <- list(returns = counts,
              xysize = xysize, xylims = xylims,
              zsize = zsize, zlims = c(zmin, zmax),
              numcols = numcols, numheights = numheights)

  class(res) <- "TLSNullModel"

  res
}


#' Main function to process scans
#'
#' This function takes scan data for a vegetation site, together with a reference
#' object giving the maximum expected number of returns per voxel, and calculates
#' the proportion of rays reflected for each voxel. The reference object is
#' generated using the function \code{get_max_returns()}. Optionally, Bayesian
#' credible intervals can be calculated for the proportion values by specifying
#' one or more probability values via the \code{probs} argument. See details below.
#'
#' Bayesian credible intervals on proportions are based on a non-informative
#' Jeffrey's prior distribution. For a given probability p, this involves taking
#' the lower (\code{(1-p)/2}) and upper (\code{(1+p)/2}) quantiles from a
#' \code{Beta(1/2 + r, 1/2 + t)} distribution, where r is the number of
#' reflected rays (returns) for the voxel, and t is the number of rays passing
#' through the voxel without being reflected. For voxels where no rays were
#' reflected, the lower bound is set to 0. For voxels where all rays were
#' reflected, the upper bound is set to 1. If a voxel was not crossed by any
#' rays, its calculated proportion will be missing (\code{NA}) and its interval
#' will also be set to \code{NA}.
#'
#' @param scandata Either a three column matrix or data frame of XYZ coordinates
#'   for return positions relative to the scanner; or a \code{LAS} object as
#'   returned by \code{lidR::readLAS}.
#'
#' @param refdata A \code{TLSNullModel} object, as returned by the function
#'   \code{get_max_returns()}, containing an array of the estimated maximum
#'   number of returns for each voxel plus the voxel and scan dimensions.
#'
#' @param ground Either a square numeric matrix, or a \code{RasterLayer} object,
#'   where values are ground elevation for each voxel. The number of rows and
#'   columns of the matrix or raster be the value of element \code{numcols} in
#'   \code{refdata}.
#'
#' @param probs Optionally, one or more probability values (0 to 1). If
#'   provided, bounds on the proportion of reflected rays are calculated for
#'   each voxel. For example, with \code{probs = c(0.5, 0.9)} lower and upper
#'   50 and 90 percent bounds will be calculated.
#'
#' @param fail.xy If \code{TRUE}, the function will stop with an error if any
#'   scan points lie outside the horizontal bounds of the reference model. If
#'   \code{FALSE} (the default), the function will issue a warning and ignore
#'   the points.
#'
#' @param fail.z If \code{TRUE} (the default), the function will stop with an
#'   error if any scan point lie above or below the vertical limits of the
#'   reference model. If \code{FALSE}, the function will issue a warning and
#'   ignore the points.
#'
#' @return A named list (object of class \code{TLSResult}) with the following
#'   elements:
#'   \describe{
#'   \item{preflected}{An array (dimensions: X, Y, Z) of the proportion of rays
#'     reflected for each voxel.}
#'   \item{bounds_XX}{If argument probs was specified, arrays for lower and
#'     upper bounds on calculated proportions will be included,
#'     e.g. setting probs = 0.9 will results in two arrays 'bounds_0.05'
#'     and 'bounds_0.95'}
#'   \item{nreflected}{Array of the number of reflected rays for each voxel.}
#'   \item{nthrough}{Array of the number of rays passing through each voxel.}
#'   \item{nblocked}{Array of the number of rays that were blocked before
#'     reaching each voxel.}
#'   \item{xysize}{The horizontal voxel width.}
#'   \item{xlims}{A vector of min and max values for X and Y ordinates.}
#'   \item{zsize}{The vertical voxel width.}
#'   \item{zlims}{A vector or min and max values for Z (height).}
#'   \item{numcols}{NUmber of voxels in each horizontal dimension.}
#'   \item{numheights}{Number of voxels in vertical dimension.}
#'   }
#'
#' @export
#'
get_prop_reflected <- function(scandata, refdata, ground,
                               probs = NULL,
                               fail.xy = FALSE,
                               fail.z = TRUE) {

  if (!inherits(refdata, "TLSNullModel")) {
    stop("refdata should be a named list of class TLSNullModel")
  }

  if (inherits(scandata, "LAS")) {
    scandata <- as.matrix(scandata@data[, c("X", "Y", "Z")])

  } else if (is.matrix(scandata) || is.data.frame(scandata)) {
    if (!ncol(scandata) == 3) {
      stop("Argument scandata should be a three column matrix or data frame")
    }
    scandata <- as.matrix(scandata)
  }

  # Check probs argument
  if (length(probs) > 0) {
    if (!is.numeric(probs) || !is.vector(probs) || !all(probs > 0 & probs < 1)) {
      stop("Argument probs should be a vector of one or more values between 0 and 1")
    }
  }

  numcols <- refdata$numcols
  xylims <- refdata$xylims
  xysize <- refdata$xysize
  zsize <- refdata$zsize

  # Check that scan points are within the horizontal bounds of the reference model
  ok <- scandata[,1] >= xylims[1] & scandata[,1] < xylims[2] &
                scandata[,2] >= xylims[1] & scandata[,2] < xylims[2]

  if (any(!ok)) {
    n <- sum(!ok)
    msg <- ifelse(n == 1, "point is outside", "points are outside")
    msg <- glue::glue("{n} {msg} the horizontal bounds of the reference model")

    if (fail.xy) {
      stop(msg)
    }

    # fail.xy is FALSE so just issue a warning and remove the points
    msg <- glue::glue("{msg} and will be ignored")
    warning(msg, immediate. = TRUE)
    scandata <- scandata[ok, ]
  }

  # Check for points beyond the vertical range of the reference model
  ok <- scandata[,3] >= refdata$zlims[1] & scandata[,3] <= refdata$zlims[2]
  if (any(!ok)) {
    n <- sum(!ok)
    msg <- ifelse(n == 1, "point is outside", "points are outside")
    msg <- glue::glue("{n} {msg} the vertical limits of the reference model")

    if (fail.z) {
      stop(msg)
    }

    # fail.z is FALSE so just issue a warning and remove the points
    msg <- glue::glue("{msg} and will be ignored")
    warning(msg, immediate. = TRUE)
    scandata <- scandata[ok, ]
  }

  # Make sure we still have some data
  if (nrow(scandata) == 0) {
    stop("No remaining scan data")
  }

  # Check ground elevation data
  if (inherits(ground, c("matrix", "RasterLayer"))) {
    if (nrow(ground) != refdata$numcols ||
        ncol(ground) != refdata$numcols) {

      stop("The matrix or RasterLayer of ground elevations should be square,\n",
           "with the number of rows and columns equal to the 'numcols' value\n",
           "in refdata")
    }

    ground <- as.matrix(ground)

  } else if (is.numeric(ground)) {
    ground <- .single_numeric(ground)
    ground <- matrix(ground, numcols, numcols)

  } else {
    stop("Argument ground should either be a square matrix with number of\n",
         "rows and columns equal to the 'numcols' value in refdata, or a\n",
         "single value for ground elevation to apply to all voxels.")
  }

  if (all(is.na(ground)))
    stop("All ground elevation values are missing.")


  if (anyNA(ground)) {
    ground <- nibble_matrix(ground)
  }

  # Determine the height range and number of voxel height levels.
  # We clamp the height limits to multiples of zsize.
  minz <- min(ground)
  minz <- (trunc(minz / zsize) - 1) * zsize
  maxz <- (trunc(max(scandata[,3]) / zsize) + 1) * zsize

  stopifnot(maxz > minz)
  numheights <- (maxz - minz) / zsize


  # Z-index offset for reference model.
  ref.zoffset <- (minz - refdata$zlims[1]) / zsize

  # calculate maximum distance of points from scanner
  maxzdist <- max(abs(c(maxz, minz)))
  maxdist <- sqrt(2*xylims[2]^2 + maxzdist^2)

  # Call Rcpp function to count the total number of rays entering each voxel,
  # the number reflected, and the number passing through.
  raycounts <- cpp_count_voxel_rays(
    scandata,
    numcols, numheights,
    xysize, xylims[1], xylims[2],
    zsize, minz, maxz,
    maxdist,
    TRUE,
    ground
  )

  dims <- c(numcols, numcols, numheights)
  totalrays <- array(raycounts$total, dim = dims)
  reflectedrays <- array(raycounts$reflected, dim = dims)
  throughrays <- array(raycounts$through, dim = dims)
  blockedrays <- array(raycounts$blocked, dim = dims)

  # To account for rays that went through a voxel without being returned,
  # we need to add these rays into the 'through' tally by comparison with
  # the null model scan performed in a closed room.
  for (yrow in 1:numcols) {
    for (xcol in 1:numcols) {
      for (zlayer in 1:numheights) {
        # Check if we have less rays in the voxel than expected (because
        # some went through). If so, increase the through count.
        expected <- refdata$returns[yrow, xcol, zlayer + ref.zoffset]
        if (totalrays[yrow, xcol, zlayer] < expected) {
          nadd <- expected - totalrays[yrow, xcol, zlayer]

          zminlimit <- (zlayer-1) * zsize + minz
          z <- zminlimit + zsize/2
          zadj <- z - ground[yrow, xcol]
          zlayer.adj <- trunc(zadj / zsize) + 1
          throughrays[yrow, xcol, zlayer.adj] <- throughrays[yrow, xcol, zlayer.adj] + nadd
        }
      }
    }
  }

  # Proportion of reflected rays per voxel
  preflected <- reflectedrays / (reflectedrays + throughrays)

  # Convert any NaN values, which occur when a voxel had no rays
  # to NA
  preflected[is.nan(preflected)] <- NA

  res <- list(preflected = preflected,
              nreflected = reflectedrays,
              nthrough = throughrays,
              nblocked = blockedrays)

  # Calculate bounds if probs were specified
  if (length(probs) > 0) {
    my <- matrix(c(reflectedrays, throughrays), ncol = 2)
    qs <- sort(unique(c((1-probs)/2, (1+probs)/2)))

    bounds <- lapply(1:length(qs), function(i) {
      x <- qbeta(qs[i], 1/2 + my[,1], 1/2 + my[,2])

      if (i <= length(probs)) { # lower bound
        x[my[,1] == 0] <- 0
      } else { # upper bound
        x[my[,2] == 0] <- 1
      }

      x <- array(x, dim = dim(reflectedrays))
      x[is.na(preflected)] <- NA
      x
    })

    names(bounds) <- paste0("bound_", qs)

    res <- c(res, bounds)
  }

  # Return results
  res <- c(res,
           list(xysize = xysize, xylims = xylims,
                zsize = zsize, zlims = c(minz, maxz),
                numcols = numcols, numheights = numheights) )

  class(res) <- "TLSResult"

  res
}

